### ISSUE TITLE
Add gh-pages branch to make "LZ4 Homepage" on GitHub

### ISSUE THREAD

**Issue by t-mat** [Cyan4973/lz4#55]
This is first approximation to solve "[Add list of LZ4 variants on Github](https://fastcompression.uservoice.com/forums/279225-general/suggestions/6886866-add-list-of-lz4-variants-on-github)".

### Demo
https://t-mat.github.io/lz4/

My fork is https://github.com/t-mat/lz4/tree/gh-pages. Please see [index.html](https://github.com/t-mat/lz4/blob/gh-pages/index.html). It is almost same as current `README.md`.

### `gh-pages` branch and GitHub pages

`gh-pages` is special branch for GitHub. You can see how `gh-pages` works by the following commands.
Read "[Creating Project Pages manually](https://help.github.com/articles/creating-project-pages-manually/)" for the details.

```
cd /your/workspace/gh-pages/
git clone https://github.com/Cyan4973/lz4.git
cd lz4
git checkout --orphan gh-pages
git rm -rf .
echo "LZ4 gh-pages testing 1,2,3" > index.html
git add index.html
git commit -a -m "First gh-pages commit"
git push origin gh-pages
```

Here, you can open https://Cyan4973.github.io/lz4/ in your browser.

After that, if you could merge my `gh-pages` branch.

### Further evolution

For simplicity, I use [Strapdown.js](http://strapdownjs.com/) for Markdown rendering. But obviously it is js-heavy page.
So for long term, it would be nice to introduce some static page generators. For exmaple [GitHub recommends Jekyll](https://help.github.com/articles/using-jekyll-with-pages/).

### See also

- [Setting up a custom domain with GitHub Pages](https://help.github.com/articles/setting-up-a-custom-domain-with-github-pages/)

**Comment by Cyan4973** [comment]
Thanks Takayuki,
this is a great introduction, with extremely well documented references.

I'll need a bit of time to properly integrate its content,
but it does answer a strong need which wasn't fullfilled since a few months.

**Comment by t-mat** [comment]
FYI : Google announced "[Bidding farewell to Google Code](http://google-opensource.blogspot.com/2015/03/farewell-to-google-code.html)"
> - March 12, 2015 - New project creation disabled.
> - August 24, 2015 - The site goes read-only. You can still checkout/view project source, issues, and wikis.
> - January 25, 2016 - The project hosting service is closed. You will be able to download a tarball of project source, issues, and wikis. These tarballs will be available throughout the rest of 2016.

**Comment by Cyan4973** [comment]
Yep, thanks, just noticed it...

I had only one more project still hosted there, mmc (morphing match chain), which I just exported to github.
Time to make github the primary source repository...

**Comment by Cyan4973** [comment]
I've started implementing a project page at :
 https://Cyan4973.github.io/lz4/

using your provided initial sample.

I'm not yet satisfied with current result. But that's nonetheless a welcomed start.

Maybe a few questions I've in mind are :
- What kind of editing tools does exist to easily create/modify simple web pages
- Is it necessary or preferable to keep such pages "markdown" compatible

**Comment by t-mat** [comment]
> What kind of editing tools does exist to easily create/modify simple web pages

I think "Static site generator" + "Your favorite text editor" is good tool for simple task nowadays. See https://www.staticgen.com/ to find popular generators.
If you've never used static site generator, try [my short introduction to Jekyll](https://gist.github.com/t-mat/b341f22184610f907068).

> Is it necessary or preferable to keep such pages "markdown" compatible

Not necessary, but since Markdown is widely supported in many template engines, it has an advantage in portability. If Markdown has not enough expressiveness, inline HTML in Markdown or raw HTML is good fallback.

For example, google/bazel's gh-pages use both form

 - https://github.com/google/bazel/blob/gh-pages/docs/install.md
 - https://github.com/google/bazel/blob/gh-pages/docs/bazel-user-manual.html

**Comment by Cyan4973** [comment]
Thanks for excellent references, Takayuki

**Comment by Cyan4973** [comment]
Just for information : 
Looking around, I found that the simple website at 
https://quixdb.github.io/squash/
was quite interesting.
Neat, simple, well, good enough for a presentation page.

looking at the gh branch, there is nothing fancy : 
https://github.com/quixdb/squash/tree/gh-pages

just a simple index.html file.

Asking the author about how he generated it, here is his answer : 

> I did.  Started with Bootstrap (<http://getbootstrap.com/>), then used the Sandstone theme from Bootswatch (<https://bootswatch.com/>).  I used Angular.js for the dynamic parts, and Highcharts for the charts.
> There are a couple places where I used inline styles (like to center the links to the CSVs), but it's pretty much stock.
> The data is all loaded dynamically from the CSVs.  All the JavaScript logic is in squash-benchmark.js.

Not sure if its answer is related to the squash homepage, or squash-benchmark instead : 
https://quixdb.github.io/squash-benchmark/

**Comment by t-mat** [comment]
Looks good :)
So, for now, it seems some investigation is needed for the following things:
 - Theme (Bootstrap)
 - Navigation bar
 - Headline box ("LZ4 is ...")
 - Dynamic chart
 - API document (Doxygen)
 - manpage (man2html)
 - Tracker (Google analytics)

Sorry for bad formatting and lack of information due to the mobile. I'll edit this post later.

**Comment by Cyan4973** [comment]
Yes, exactly,
well, maybe we can keep Dynamic chart for later, but the rest looks very interesting.

Don't worry about your formatting, it's perfect.
Ah, the beauty of markdown...

I'm in the process of converting LZ4 doc to Markdown format by the way.

**Comment by t-mat** [comment]
I've pushed new demo : https://t-mat.github.io/lz4/test.html

Since http://lz4.info/ is online, I created test.html for mergeable testing file. It implements the following things :

 - [X] Theme (Bootstrap)
 - [X] Navigation bar
 - [X] Headline box ("LZ4 is ...")

My commits are : https://github.com/t-mat/lz4/commits/gh-pages , 1176bf32d4d0f98b00a6dae2438c22e49a2d9219 .. cb75b3d46e8ef6e0f908bcc778d6f45fe278ebeb

**Comment by Cyan4973** [comment]
wow, this is such a great design !
I really love your new homepage proposal Takayuki !

**Comment by t-mat** [comment]
Add basic [Google Charts](https://google-developers.appspot.com/chart/) function to [test.html](https://t-mat.github.io/lz4/test.html). You can see charts at bottom of the page and https://github.com/t-mat/lz4/tree/gh-pages/csv for its source CSV.
If you are going to play with Google Charts, [Chart Gallery](https://developers.google.com/chart/interactive/docs/gallery) is good introduction.

**Comment by Cyan4973** [comment]
I just imported your updated lz4's homepage version (called "test.html")
and I must say that I'm highly impressed.
The rendering is exceptional, beyond my best expectations.

It provides an excellent ground for new presentation ideas too.
I don't know when I'll find time to work on it, but it sure is an excellent direction you've already provided to it.

Best Regards

**Comment by Cyan4973** [comment]
Finally adopted your latest proposal as the new LZ4 homepage,
accessible at http://www.lz4.info

**Comment by Cyan4973** [comment]
Hi @t-mat 

I'm currently trying to update the LZ4 homepage.
The new format you contributed is absolutely excellent, so these are just minor updates.

Among the little ones I'm failing is the integration of LZ4 documentation (block & frame) in Markdown format. The files are called `lz4_Block_format.md` and `lz4_Frame_format.md`.
The `.md` extension allows these to be automatically converted on Github.
See for example [block](https://github.com/Cyan4973/lz4/blob/dev/lz4_Block_format.md) or [frame](https://github.com/Cyan4973/lz4/blob/dev/lz4_Frame_format.md) links.

No such luck on lz4 project page. `.md` files are just considered _downloadable_ content.
Converting the extension to `.txt` just makes it displayed as a regular text file.

I guess I could build a page from `index.html`, by copy/pasting most header/footer elements and just replacing the central body with the content of a `.md` file.
But it somehow misses the point.
I would like, if that is possible, to just drag & drop the documentation, unmodified, into the `gh` branch. This way, when there are modifications, I don't need to modify multiple files.

I was hoping it could be possible to "include" the content of a `.md` file, but failed to find a way. Do you have any idea ?

If no way can be found, I'm thinking that an alternative could be to link directly to the github hosted and parsed files.

**Comment by t-mat** [comment]
Hi,

> I was hoping it could be possible to "include" the content of a .md file, but failed to find a way. Do you have any idea ?

The following experimental HTML file "includes" external Markdown file :

 - https://t-mat.github.io/lz4/lz4_Block_format.html

### Related commits

 - https://github.com/t-mat/lz4/commit/56216ac16079a67f650b34ad94fd738f16da68bf
 - https://github.com/t-mat/lz4/commit/8c4989e2b9fa5434259029c95e92c0ed40ec000a

### Usage

 - Add [js/strapdown-hook.js](https://github.com/t-mat/lz4/commit/56216ac16079a67f650b34ad94fd738f16da68bf) to gh-pages.
 - Add ["src" attribute](https://github.com/t-mat/lz4/blob/gh-pages/lz4_Block_format.html#L61) to `<xmp>` tag.

```HTML
<xmp theme="sandstone" style="display:none;" src="lz4_Block_format.md"></xmp>
```

 - Change strapdown `script` tag's `src` to [`js/strapdown-hook.js`](https://github.com/t-mat/lz4/blob/gh-pages/lz4_Block_format.html#L70) instead of `strapdown/v/0.2/strapdown.js`.

### Possible Problems

 - Copy pastes :weary:
 - Blank page will be shown in non-JS environment.

----

> No such luck on lz4 project page. .md files are just considered downloadable content.

Markdown (`.md`) file needs .md to HTML converter.  But since GitHub Pages accepts only static HTML and resources, we can't execute any server-side scripts/parsers.

**Comment by Cyan4973** [comment]
Thanks for your precious help Takayuki

Your proposed modifications seem to work as intended,
with potentially the 2 following minor issues : 

  1. Parsing is slightly different from Github. 
      Look typically at the paragraph which provides some length examples. 
      Lists have disappeared. Content is simply clogged together.
      I suspect that adding a blank line before the first '-' will do the break,
      but then, it would change a bit the layout
      and the whole point is to have a document which can be read as a "raw" text file
      and can be safely interpreted by a MarkDown compatible parser

  2. lz4_Block_format.html doesn't work on my test system, which is Firefox 37 for Linux Mint 17.
      Top band is different (white), and page is blank.
      That may be what you call "non-JS environment".
      That's a bit strange though, as the first page (lz4 homepage) works correctly.

Regards

**Comment by t-mat** [comment]
> 1. Parsing is slightly different from Github. 

Yes.  [Lack of standard syntax](http://en.wikipedia.org/wiki/Markdown#Standardization) is a biggest problem of Markdown.  Also since Strapdown.js uses (includes) relatively old version of JS Markdown parser which is called "[marked](https://github.com/chjj/marked)", it does not support relatively newer variants such as GFM and commonmark.

Unfortunately, I think there is no definitive way to resolve this problem.  [Babelmark 2](http://johnmacfarlane.net/babelmark2/?normalize=1&text=abc%0A-+def%0A) will help to find incompatibility, but it does not solve the problem.


> 2. lz4_Block_format.html doesn't work on my test system,
> (but works fine on http://www.lz4.info)
> That's a bit strange though, as the first page (lz4 homepage) works correctly.

I'm not sure about your test system, but it seems somekind of URL problem.  I'll investigate it soon.  But you can also test your system with the following commands :

```
cd /path/to/lz4-gh-pages/

# make sure there is index.html
cat index.html

python3 -m http.server 8080
# for python2
# python2 -m SimpleHTTPServer 8080

xdg-open http://localhost:8080/lz4_Block_format.html
```

Also [Firebug](http://getfirebug.com/)'s "Net" tab helps to find resource loading problems.

**Comment by Cyan4973** [comment]
> it seems somekind of URL problem. 

Thanks for the hint. Found it.

Changed the following lines from : 
````
    <!-- jQuery (necessary for Bootstrap's JavaScript plugins) -->
    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.2/jquery.min.js"></script>
    <!-- Include all compiled plugins (below), or include individual files as needed -->
    <script src="//maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
````

to : 
````
    <!-- jQuery (necessary for Bootstrap's JavaScript plugins) -->
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.2/jquery.min.js"></script>
    <!-- Include all compiled plugins (below), or include individual files as needed -->
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
````

And now it's fine.

Don't know why the first version worked when it was hosted on github. Maybe because https:// is implied.

**Comment by t-mat** [comment]
> Don't know why the first version worked when it was hosted on github.

I suppose you have tested yourfiles by `file://` protocol.  Since it's 'protocol-less' URL, [`file://` is not working properly](http://stackoverflow.com/a/4832046/2132223).  I also noticed 'protocol-less' URLs are [bad habit nowadays](http://www.paulirish.com/2010/the-protocol-relative-url/).  So your change is correct and desirable :+1:

**Comment by Cyan4973** [comment]
Excellent reference. Thanks Takayuki

**Comment by Cyan4973** [comment]
I do have 2 minor questions in mind since yesterday's update : 

> Lack of standard syntax is a biggest problem of Markdown. 

I was reading the documentation of [Sundown](https://github.com/vmg/sundown),
which states : 

> Sundown passes out of the box the official Markdown v1.0.0 and v1.0.3 test suites

So it seems there is some kind of "official Markdown test suite".
Not that it will solve everything, but it should at least strengthen conformance on the most important feature set.

The second question is about the loading of JS elements from external source,
most importantly https://maxcdn.bootstrapcdn.com/ and https://ajax.googleapis.com/.

Is it standard practice to depend on such external source ? (as opposed to having the files into one's own directory structure ?)

I'm a bit surprised, because there is no guarantee that it will always work, and it makes these providers an ideal target to infect a huge number of dependent sites. But then, I have no experience on deploying JS-enabled website, so I simply don't know if it is standard practice, or just convenient.

Regards

**Comment by t-mat** [comment]
> So it seems there is some kind of "official Markdown test suite".

I don't know anything about "official" test suites.  I think Markdown is (some kind of) abandonware without major successor and authority.

> Is it standard practice to depend on such external source ? (as opposed to having the files into one's own directory structure ?)

Yes.  There are offical (sponsored) CDN and hosted CDN by major sites such as Google, Microsoft, etc.

 - [Bootstrap CDN](http://getbootstrap.com/getting-started/#download-cdn)
 - [JQuery CDN](https://code.jquery.com/)
 - [Hosted Libraries @ Google Developers](https://developers.google.com/speed/libraries/)

See the following Q&A for the details :

 - [Why should I use Google's CDN for jQuery? @ SO](http://stackoverflow.com/questions/2180391/why-should-i-use-googles-cdn-for-jquery)

> I'm a bit surprised, because there is no guarantee that it will always work, and it makes these providers an ideal target to infect a huge number of dependent sites.

Yes.  So there is no problem with having your own copy.

It's like static / dynamic link problem.  Each method works fine and has its own pros/cons.  But ordinal user favor automatic and maintenance-free method.

**Comment by Cyan4973** [comment]
Thanks for reference Takayuki. It's an excellent read, as usual.
